syntax = "proto3";

package multicast;

import "google/protobuf/empty.proto";

// Possible kinds of messages.
enum Kind {
  DEFAULT = 0;
  MESSAGE = 1;
  ACK = 2;
}

// The multicast message.
message MulticastMessage {
  // The Lamport clock of the message.
  uint32 lamport_clock = 1;

  // The sender of the message.
  string sender = 2;

  // The kind of the message.
  Kind kind = 3;

  // The message content.
  optional string content = 4;
}

// The multicast service.
service MulticastService {
  // Receives a multicast message.
  rpc ReceiveMessage(MulticastMessage) returns (google.protobuf.Empty);

  // Test multiple connections
  rpc Test(MulticastMessage) returns (MulticastMessage);
}

